mzLib/FlashLFQ/PEP/PEPValueAnalysisGeneric.cs

```diff
@@ -0,0 +1,670 @@
+        public List<int>[] Get_Donor_Group_Indices(List<DonorGroup> donors, int numGroups)
+        {
+            List<int>[] groupsOfIndices = new List<int>[numGroups];
+            for (int i = 0; i < numGroups; i++)
+            {
+                groupsOfIndices[i] = new List<int>();
+            }
+
+            int myIndex = 0;
+
+            while (myIndex < donors.Count)
+            {
+                int subIndex = 0;
+                while (subIndex < numGroups && myIndex < donors.Count)
+                {
+                    groupsOfIndices[subIndex].Add(myIndex);
+
+                    subIndex++;
+                    myIndex++;
+                }
+            }
+
+            // We're trying to equalize the number of targets and decoys in each group
+            // This is going to be extremely cursed
+            for(int i = 0; i < numGroups-1; i++)
+            {
+                int targets1 = 0;
+                int targets2 = 0;
+                int decoys1 = 0;
+                int decoys2 = 0;
+                foreach (int index in groupsOfIndices[i])
+                {
+                    targets1 += donors[index].TargetAcceptors.Count(peak => peak.MbrScore > PipScoreCutoff);
+                    decoys1 += donors[index].DecoyAcceptors.Count;
+                }
+                foreach (int index in groupsOfIndices[i + 1])
+                {
+                    targets2 += donors[index].TargetAcceptors.Count(peak => peak.MbrScore > PipScoreCutoff);
+                    decoys2 += donors[index].DecoyAcceptors.Count;
+                }
+
+                bool stuck = false;
+                int minIndex = groupsOfIndices[i].Min();
+
+                int totalIterations = 0;
+
+                int targetSurplus = targets1 - targets2;
+                int decoySurplus = decoys1 - decoys2;
+                while((Math.Abs(targetSurplus) > 1 | Math.Abs(decoySurplus) > 1) 
+                    && !stuck)
+                {
+                    if(totalIterations > 50)
+                    {
+                        break;
+                    }
+                    totalIterations++;
+                    bool swapped = false;
+
+
+                    // start from the bottom of group 1, trying to swap peaks.
+                    // Lets do targets first
+                    int innerIterations = 0;
+                    while (Math.Abs(targetSurplus) > 1 & !stuck & innerIterations < 25)
+                    {
+                        innerIterations++;
+                        swapped = false;
+                        foreach (int index in groupsOfIndices[i].OrderByDescending(idx => idx))
+                        {
+                            if (targetSurplus > 0)
+                            {
+                                if (donors[index].TargetAcceptors.Count(peak => peak.MbrScore > PipScoreCutoff) > 1)  // if we have a positive number of targets
+                                {
```

Comment:
There are a bunch of conditionals and indices being thrown around here without test coverage for a crucial part of the code. Could this be adjusted or tested?